implementations/rust/ockam/ockam_api/src/kafka/portal_worker.rs

```diff
@@ -0,0 +1,472 @@
+/// │        │◄────────────┤ Response│◄────────────┤        │
+/// └────────┘             └─────────┘             └────────┘
+///```
+pub(crate) struct KafkaPortalWorker {
+    //the instance of worker managing the opposite: request or response
+    //the first one to receive the disconnect message will stop both workers
+    other_worker_address: Address,
+    reader: KafkaDecoder,
+    writer: KafkaEncoder,
+    receiving: Receiving,
+    shared_protocol_state: ProtocolState,
+    inlet_map: KafkaInletMap,
+    disconnect_received: Arc<AtomicBool>,
+}
+
+#[ockam::worker]
+impl Worker for KafkaPortalWorker {
+    type Message = PortalMessage;
+    type Context = Context;
+
+    async fn handle_message(
+        &mut self,
+        context: &mut Self::Context,
+        routed_message: Routed<Self::Message>,
+    ) -> ockam::Result<()> {
+        let onward_route = routed_message.onward_route();
+        let return_route = routed_message.return_route();
+        let local_info = routed_message.local_message().local_info().to_vec();
+        let portal_message = routed_message.as_body();
+
+        match portal_message {
+            PortalMessage::Payload(message) => {
+                let result = self
+                    .intercept_and_transform_messages(context, message)
+                    .await;
+
+                match result {
+                    Ok(maybe_kafka_message) => {
+                        if let Some(encoded_message) = maybe_kafka_message {
+                            self.split_and_send(
+                                context,
+                                onward_route,
+                                return_route,
+                                encoded_message,
+                                local_info.as_slice(),
+                            )
+                            .await?;
+                        }
+                    }
+                    Err(cause) => {
+                        return match cause {
+                            InterceptError::Io(cause) => {
+                                Err(Error::new(Origin::Transport, Kind::Io, cause))
+                            }
+                            InterceptError::Ockam(error) => Err(error),
+                        };
+                    }
+                }
+            }
+            PortalMessage::Disconnect => {
+                self.forward(context, routed_message).await?;
+
+                //the first one to receive disconnect and to swap the atomic will
+                //stop both workers
+                let disconnect_received = self.disconnect_received.swap(true, Ordering::SeqCst);
+                if !disconnect_received {
+                    context
+                        .stop_worker(self.other_worker_address.clone())
+                        .await?;
+                    context.stop_worker(context.address()).await?;
+                }
+            }
+            _ => self.forward(context, routed_message).await?,
```

Comment:
Could we please make other cases explicit, so that it requires update if we change the `PortalMessage`